### DISCUSSION TITLE
Future of `--experimental_sibling_repository_layout`?

### DISCUSSION THREAD

**Discussion by lberki** [bazelbuild/bazel#20500] [General]
We have an experimental flag that costs quite a bit to maintain: `--experimental_sibling_repository_layout`.

It makes external repositories appear under `../$REPO` instead of  `external/$REPO` in the execroo and runfiles trees, which has two benefits:
1. It frees up the name `external/`. Otherwise, Bazel doesn't support projects that have this directory (which is kind of embarrassing)
1. It puts external repositories and the main repository on equal footing, which is more sound in principle.

As it stands, "it's a forever experimental" flag: we have it implemented, we pay the maintenance price, but we have not committed to migrating to it, which is the worst of all worlds: our users in general don't derive the above benefits (since they presumably don't opt in) and we pay the maintenance price.

I think the best plan is to flip it in Bazel 8, but this is mostly based on feelings ((1) is quite painful). However, it may be a costly migration, so I could be convinced to delete the flag and give up on the whole things.

@Wyverald @meisterT @comius do you have any thoughts?

**Comment by tpudlik** [comment]
Putting external repositories and the main repository on an equal footing is not necessarily a benefit. For example, in Pigweed we use the fact that external repos are under `external/` to silence compiler warnings from external code using `--per_file_copt=external/.*@-w` [in our bazelrc](https://cs.opensource.google/pigweed/pigweed/+/master:.bazelrc;l=15-20;drc=e91b4e2be600f812824a1b6235f534cbf410b3dc). This is essential if you want your toolchain to apply warnings of any strictness, since without this feature you'll be seeing warnings from your dependencies.

Incidentally, this should ideally be done through toolchain configuration rather than `.bazelrc`, but there's no support for that at present (https://github.com/bazelbuild/rules_cc/issues/209).

**Reply by tjgq** [reply]
Isn't `--per_file_copt` documented to take a label regex as an argument, not a path regex? In my mind, something like `--per_file_copt=^@.+//.*@-w` **ought** to work (the regex part being `^@.+//.*` to match labels in any non-main repository), but we fail to parse it correctly because we split at the first `@`. (I think we recently fixed a similar issue with a different flag, but I can't find it now.)

**Reply by tpudlik** [reply]
The [documentation](https://bazel.build/reference/command-line-reference#flag--host_per_file_copt) is not very clear. At the very top is says, "--host_per_file_copt=<a comma-separated list of regex expressions with prefix '-' specifying excluded **paths** followed by an @ and a comma separated list of options>" (emphasis added). Since it explicitly refers to paths, it sounds like a path regex. But later, the example provided is `--host_per_file_copt=//foo/.*\.cc,-//foo/bar\.cc@-O0`, which looks like a label regex. Actually, if this is a label regex, I'm surprised `--per_file_copt=external/.*@-w` works at all (the labels don't include the `external`, do they?).

If `--per_file_copt=^@.+//.*@-w` did work, that would address my concern! (But is it certain that files in my own repository are never referred to using a label starting with their canonical repository name?)

**Reply by Wyverald** [reply]
I've only ever seen it used with execroot paths. Don't think labels would work.

**Reply by Wyverald** [reply]
Ah, looking at the [code](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java;l=1754-1763;drc=51f71f1a0f12a597e6166e43907e9f2d0de0222e), it seems that the regex is used against *both* the label and the artifact path. The problem is (IIUC) that some artifacts might not have labels at all (like certain generated sources?)

But obviously the other problem is the `@` being used as the separator between regex and options.

**Reply by Wyverald** [reply]
> But is it certain that files in my own repository are never referred to using a label starting with their canonical repository name?

This is true as of today -- some file in your own repo would appear as `//pkg:file.cc`, with no starting `@`s at all. This is arguably a quirk of the `Label.toString()` behavior.

**Reply by illicitonion** [reply]
Reference to issue about this: https://github.com/bazelbuild/bazel/issues/19665

**Comment by tjgq** [comment]
I agree that having two code paths is very unfortunate from a testing surface standpoint. I share the sentiment around the sibling layout being cleaner, but pragmatically speaking, if we can't find the time/energy to do the migration soon™, we should just cut our losses and accept the ugly/embarrassing API.

On that note, it would help to know how many people are actively using the sibling layout and haven't found any bugs in Bazel around it. (Otherwise, this might not be "just" a matter of flipping the flag and fixing downstream projects.)

**Reply by jin** [reply]
+1. This is a nearly decade-old design decision that's caused delays in Bazel adoptions (most notably in Android, when we had to _implement_ this). The workaround is ugly (Bazel doesn't work in _your_ source tree, so change _your_ source tree), so IMO there's value in getting it over the hump, leveraging on major version changes.

**Reply by fmeum** [reply]
@tjgq https://github.com/bazelbuild/bazel/discussions/20500#discussioncomment-7840203 is the one example of a bug in the sibling layout I am aware of.

**Comment by Wyverald** [comment]
> It makes external repositories appear under `../$REPO` instead of `external/$REPO` in the execroo and runfiles trees, which has two benefits:

Does this actually affect the runfiles tree? I thought runfiles trees were already organized as `$runfilesRoot/$repoName` without any nesting under `$runfilesRoot/_main/external/$repoName`.

> do you have any thoughts?

I'm mostly worried about @tpudlik's comment: IMO we definitely need a solution to the `--per_file_copts` problem before we can even think about flipping this.

**Reply by lberki** [reply]
When the flag flipped, the paths of files in external repositories starts with `../` instead of `external/` so the regex could be changed accordingly (or even to accommodate both alternatives).

So it looks like `--per_file_copt` isn't the thing this decision should hinge upon.

**Comment by fmeum** [comment]
I would also welcome the flag flip. Another benefit that hasn't been mentioned yet is that it aligns the directory structure of input files with that of runfiles, which would have helped me a lot when I initially got into rules authoring.

That said, if we want to flip this in Bazel 8, I think that we need to resolve the following issues soon (i.e., in time for Bazel 7.1 so that ruleset authors have ample time to prepare the migration):

1. https://groups.google.com/g/bazel-discuss/c/eyjCTJl3OHA: Even with `--experimental_sibling_repository_layout`, `bazel-out` for an external repo is still rooted under the main repo (e.g. `__main__/bazel-out/other_repo/k8-fastbuild/bin/pkg/gen.txt`). This voids most of the gains in conceptual simplicity and can also break builds if `other_repo` happens to be called `k8-fastbuild`.
2. https://github.com/bazelbuild/bazel/issues/16472: Without a `local_includes` attribute on `cc_library`, users currently tend to hardcode `external/...` in BUILD files. There are unofficial workarounds (see https://github.com/bazelbuild/bazel/issues/15553), so an alternative could be to standardize on one and make it available in `rules_cc`. This would be less discoverable though.

With these issues addressed and with buy-in from the Rules Authors SIG to address ruleset incompatibilities, this migration should be doable.

**Reply by lberki** [reply]
re (1): it was an explicit decision to put the `bazel-out` of external repositories under the cwd. If I recall correctly, one of the arguments for it was that it turned out that a single "output directory" was much easier to handle in the code than one for each external repository and it's also convenient that all output files are accessible under a single convenience symlink. It does open us up to name clashes like the way you mention and I'm still of the opinion that if we want to flip this flag, we might revisit this decision.

re (2), I think it's indicative of the long-tail issues we'd be facing if we committed to this flag flip.

**Comment by lberki** [comment]
I sent out https://github.com/bazelbuild/bazel/pull/21785 to disable the flag :(

**Comment by lberki** [comment]
Update: as a last-ditch effort to rescue this flag flip, I checked the downstream project pipeline and it looked like there was only a very manageable number of breakages. In light of this, @comius @meteorcloudy and @meisterT kindly agreed to push the migration forward, so instead of removing this flag, we are going to flip it. Yay!

**Reply by fmeum** [reply]
Sounds great! Happy to help with the migration as well. What's the plan for dealing with https://github.com/bazelbuild/bazel/discussions/20500#discussioncomment-7842090 though?

**Reply by lberki** [reply]
I'm afraid the plan I have in my mind is "deal with it later". It would require a quite serious reshuffling of both the output directory layout and the code that creates it and I'd much rather do it separately, even if that means two changes to the output directory structure instead of one.

I have another theoretically annoying issue, which is that with sibling repository layout, the cwd really should not be `execroot/<main repo>`, but simply `execroot` so that all repositories are equal, but that's *another* big change and I have no idea if its benefits would outweigh its costs or not.

**Reply by fmeum** [reply]
I'm worried about the cost of doing all these migrations separately, as well as the risk of getting stuck somewhere in the middle. For example, some runfiles libraries would need to support every single format (they derive the canonical repo name of a file from its execpath) and this can get out of hand if we keep changing it. Happy to look into more compatible alternatives, but that again would require a better understanding of future migrations.

**Comment by lberki** [comment]
A few months later, it seems that we are switching back to "this flag should be condemned":

https://github.com/bazelbuild/bazel/issues/23576#issuecomment-2345688136

Please speak up if you think that's a bad idea -- given that we haven't made progress in the legwork to clean up the flag, that it has a large potential for harm, that no one benefits from it too much and that it's arguably the wrong layout anyway, it's hard to imagine an argument for it...

**Reply by AlexK-BD** [reply]
`--legacy_external_runfiles=false` produces the sibling layout for tests and such -- wouldn't it be good if the layout was consistent for execroot and runfiles?

**Reply by meteorcloudy** [reply]
See https://github.com/bazelbuild/bazel/issues/23576#issuecomment-2345783288

And the fact [flipping --legacy_external_runfiles to false](https://github.com/bazelbuild/bazel/issues/23574) didn't cause much breakages indicates most users are already depending on the new layout.

**Reply by AlexK-BD** [reply]
I'm working on adopting `--nolegacy_external_runfiles` at the moment! One example consistency problem I've run into in the process:

- We have a `license_files_list` rule that uses an aspect to collect the `file.path` for license files in its transitive dependencies. `file.path` produces paths with `external/whatever` for files that are in other workspaces. This produces a generated file, `license_list.txt` for example.
- There are things that want to read `license_list.txt` at runtime and then access the licenses listed therein. But, the licenses are no longer found under `external`, they're in the "sibling" layout.

Is there a better thing to use than `file.path` to get the location that a file will have in a runfiles directory?

**Comment by lberki** [comment]
Almost two years later, I haven't heard any voices defending `--experimental_sibling_repository_layout`. Accordingly, I'm planning to approve https://github.com/bazelbuild/bazel/pull/30586 that @fmeum kindly put together tomorrow or the day after that (depending on my travel schedule).

If you feel strongly about this flag, the time to speak up is now.

**Comment by lberki** [comment]
It's over, Frodo. https://github.com/bazelbuild/bazel/commit/ce2477d7eb436425c4d056c7fe6ff13e258ae267 makes `--experimental_sibling_repository_layout` a no-op.

It was a hopeful experiment that did not live up to its promise.
